File: TwitterStreaming.scala
import org.apache.spark.rdd.RDD
import org.apache.spark.streaming.twitter._
import org.apache.spark.streaming.{Seconds, StreamingContext}
import org.apache.spark.{SparkConf, SparkContext}
import twitter4j.Status
import scala.collection.mutable.ListBuffer
import scala.util.Random
import scala.collection.mutable.Map

object TwitterStreaming {

  var N = 0
  val S = 100
  var sampleTweets = new ListBuffer[Status]()
  var sampleTweetLength = 0
  var hashtagsDict = Map[String, Int]()

  def ProcessTweetBatches(tweetBatches : RDD[Status]) : Unit = {

    val tweets = tweetBatches.collect()
    for(status <- tweets){

      if(N < S){
        sampleTweets.append(status)
        sampleTweetLength = sampleTweetLength + status.getText().length

        val hashTags = status.getHashtagEntities().map(_.getText)
        for(tag <- hashTags){
          if(hashtagsDict.contains(tag)){
            hashtagsDict(tag) += 1
          }
          else{
            hashtagsDict(tag) = 1
          }
        }
      }

      else{
        val j = Random.nextInt(N)
        if(j < S){
          val tweetToBeRemoved = sampleTweets(j)
          sampleTweets(j) = status
          sampleTweetLength = sampleTweetLength + status.getText().length - tweetToBeRemoved.getText().length

          // Remove old hashtags
          val hashTags = tweetToBeRemoved.getHashtagEntities().map(_.getText)
          for(tag <- hashTags){
            hashtagsDict(tag) -= 1
          }

          // Add new hashtags
          val newHashTags = status.getHashtagEntities().map(_.getText)
          for(tag <- newHashTags){
            if(hashtagsDict.contains(tag)){
              hashtagsDict(tag) += 1
            }
            else{
              hashtagsDict(tag) = 1
            }
          }

          val topTags = hashtagsDict.toSeq.sortWith(_._2 > _._2)
          val size = topTags.size.min(5)

          println("The number of the twitter from beginning: " + (N + 1))
          println("Top 5 hot hashtags:")

          for(i <- 0 until size){
            if(topTags(i)._2 != 0){
              println(topTags(i)._1 + ":" + topTags(i)._2)
            }
          }

          println("The average length of the twitter is: " +  sampleTweetLength/(S.toFloat))
          println("\n\n")
        }
      }
      N = N + 1
    }
  }

  def main(args: Array[String]): Unit = {

    val spark = new SparkConf()
    spark.setAppName("TwitterStreaming")
    spark.setMaster("local[*]")
    val sc = new SparkContext(spark)
    sc.setLogLevel(logLevel = "OFF")

    val consumer_key = "<ENTER_YOUR_OWN>"
    val consumer_secret = "<ENTER_YOUR_OWN>"
    val access_token = "<ENTER_YOUR_OWN>"
    val access_token_secret = "<ENTER_YOUR_OWN>"
    System.setProperty("twitter4j.oauth.consumerKey", consumer_key)
    System.setProperty("twitter4j.oauth.consumerSecret", consumer_secret)
    System.setProperty("twitter4j.oauth.accessToken", access_token)
    System.setProperty("twitter4j.oauth.accessTokenSecret", access_token_secret)

    val ssc = new StreamingContext(sc, Seconds(10))
    val tweets = TwitterUtils.createStream(ssc, None, Array("Data"))
    tweets.foreachRDD(tweetBatches => ProcessTweetBatches(tweetBatches))
    ssc.start()
    ssc.awaitTermination()
  }
}

File: TwitterStreaming.py
import tweepy
import random
import operator

N = 0
S = 100
sampleTweets = []
sampleTweetLength = 0
hashtagsDict = {}

class MyStreamListener(tweepy.StreamListener):
	
	def on_status(self, status):
		global N
		global S
		global sampleTweets
		global sampleTweetLength
		global hashtagsDict
		
		if N < S:
			sampleTweets.append(status)
			sampleTweetLength += len(status.text)

			hashtags = status.entities['hashtags']
			for hashtag in hashtags:
				tag = hashtag['text']
				if tag in hashtagsDict:
					hashtagsDict[tag] += 1
				else:
					hashtagsDict[tag] = 1


		else:
			j = random.randrange(0, N)
			if j < S:
				tweetToBeRemoved = sampleTweets[j]
				sampleTweets[j] = status 
				sampleTweetLength = sampleTweetLength + len(status.text) - len(tweetToBeRemoved.text)

				# Remove old hashtags
				hashtags = tweetToBeRemoved.entities['hashtags']
				for hashtag in hashtags:
					tag = hashtag['text']
					hashtagsDict[tag] -= 1

				# Add new hashtags
				hashtags = status.entities['hashtags']
				for hashtag in hashtags:
					tag = hashtag['text']
					if tag in hashtagsDict:
						hashtagsDict[tag] += 1
					else:
						hashtagsDict[tag] = 1

				# Top 5 hashtags
				topTags = sorted(hashtagsDict.items(), key=operator.itemgetter(1), reverse=True)[:5]

				print "The number of the twitter from beginning:", N + 1
				print "The top 5 hot hashtags:"
				for topTag in topTags:
					if topTag[1]==0:
						continue
					print topTag[0].encode('utf-8') + ":" + str(topTag[1])

				print "The average length of the twitter is:", sampleTweetLength/float(S)
				print "\n\n"

		N = N + 1

	def on_error(self, status_code):
		if status_code == 420:
			#returning False in on_data disconnects the stream
			return False

consumer_key = "<ENTER_YOUR_OWN>"
consumer_secret = "<ENTER_YOUR_OWN>"
access_token = "<ENTER_YOUR_OWN>"
access_token_secret = "<ENTER_YOUR_OWN>"

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_token_secret)
api = tweepy.API(auth)

myStreamListener = MyStreamListener()
myStream = tweepy.Stream(auth = api.auth, listener=myStreamListener)

#myStream.sample(languages=['en'])
myStream.filter(track=['Data'])